docs/proposals/vip-4.md

```diff
@@ -0,0 +1,273 @@
+|---------------------------------------------:|:----------:|:---:|:---:|:---:|
+|                         `validateHookConfig` |   ➡️ ☠️    |  ❌  |  ✅  |  ❌  |
+|                          `preStartOfPushJob` |   ➡️ ☠️    |  ❌  |  ✅  |  ✅  |
+|                         `postStartOfPushJob` |    None    |  ❌  |  ✅  |  ✅  |
+|                      `preSchemaRegistration` |   ➡️ ☠️    |  ❌  |  ✅  | 1️⃣ |
+|                     `postSchemaRegistration` |    None    |  ✅  |  ✅  | 1️⃣ |
+|                    `preStoreVersionCreation` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                   `postStoreVersionCreation` |    None    |  ✅  |  ✅  |  ✅  |
+|  `preStartOfStoreVersionIngestionForDaVinci` | ➡️ ☠️ ✋ ↩️ | 2️⃣ | 2️⃣ | 2️⃣ |
+| `postStartOfStoreVersionIngestionForDaVinci` | ➡️ ☠️ ✋ ↩️ | 2️⃣ | 2️⃣ | 2️⃣ |
+|          `postStoreVersionLeaderReplication` |    None    |  ✅  |  ❌  |  ❌  |
+|                        `preStoreVersionSwap` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                       `postStoreVersionSwap` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                            `preEndOfPushJob` | ➡️ ☠️ ✋ ↩️ |  ✅  |  ✅  |  ✅  |
+|                           `postEndOfPushJob` |    None    |  ✅  |  ✅  |  ✅  |
+
+**Legend:**
+
+* The Actions column represent which control mechanism is available to each hook:
+  * ➡️ Proceed: Move forward with this step.
+  * ☠️ Abort: Cancel this step (and as a consequence, short-circuit any future step that would come after this one).
+  * ✋ Wait: Let the hooks framework re-run this step later (i.e. 1 minute later, by default).
+  * ↩️ Rollback: Let the store go back to the store-version it had prior to beginning the push job (in all regions). 
+
+* The last three columns are the feasibility of implementing this hook in a given component:
+  * **CC** Child Controller.
+  * **PC** Parent Controller.
+  * **VPJ** Venice Push Job.
+  * ✅ It is feasible to implement this hook within this component (without unreasonable complexity).
+  * ❌ It is NOT feasible to implement this hook within this component (without unreasonable complexity).
+  * 1️⃣ Schema registration hooks in push jobs could only be invoked in cases where auto-registration is enabled and 
+  the new schema originates from the push job itself, whereas schema registrations which are performed directly on the 
+  controller could not trigger the hook.
+  * 2️⃣ The hook for the start of ingestion for Da Vinci Clients is tricky for a few reasons. The start of ingestion is 
+  controlled by updating the Meta Store, which is a non-replicated system store updated by child controllers, so those
+  must be involved (either by running the hook there in the first place, or by having some mechanism that enables the
+  parent or VPJ to inform the child controllers of when the system store is eligible for getting updated, such as by
+  adding a new field to the `AddVersion` admin channel command). However, see Rollback Support below for why running 
```

Comment:
Today, DaVinci doesn't support quick rollback as the backup version will be deleted immediately right after the version swap, so are we considering to support fast rollback in DaVinci by keeping two versions in the disk of DaVinci nodes?

Reply:
Fast rollback in DVC is orthogonal to this project (out of scope).

Reply:
I understand it is orthogonal, and I was wondering whether we should speed it up because of this project since DVC rollback could take 10+ mins depending on the store size and during that period, there will be prod impact even inside a single colo.